src/utils/BaseHook.sol

```diff
@@ -0,0 +1,230 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {Hooks} from "@uniswap/v4-core/src/libraries/Hooks.sol";
+import {IPoolManager} from "@uniswap/v4-core/src/interfaces/IPoolManager.sol";
+import {IHooks} from "@uniswap/v4-core/src/interfaces/IHooks.sol";
+import {BalanceDelta} from "@uniswap/v4-core/src/types/BalanceDelta.sol";
+import {PoolKey} from "@uniswap/v4-core/src/types/PoolKey.sol";
+import {BeforeSwapDelta} from "@uniswap/v4-core/src/types/BeforeSwapDelta.sol";
+import {ImmutableState} from "../base/ImmutableState.sol";
+
+/// @title Base Hook
+/// @notice abstract contract for hook implementations
+abstract contract BaseHook is IHooks, ImmutableState {
+    error NotSelf();
+    error InvalidPool();
+    error HookNotImplemented();
+
+    constructor(IPoolManager _manager) ImmutableState(_manager) {
+        validateHookAddress(this);
+    }
+
+    /// @dev Only this address may call this function
+    modifier selfOnly() {
+        if (msg.sender != address(this)) revert NotSelf();
+        _;
+    }
+
+    /// @dev Only pools with hooks set to this contract may call this function
+    modifier onlyValidPools(IHooks hooks) {
```

Comment:
when is this used?

Reply:
could these just be implemented by the hooks that extend from this contract themselves?

Reply:
removed unused modifiers